cumulusci/tasks/bulkdata/step.py

```diff
@@ -0,0 +1,246 @@
+            for state_message in messages:
+                self.logger.error(f"Batch failure message: {state_message}")
+
+        return result
+
+
+class Step:
+    def __init__(self, sobject, operation, api_options, context):
+        self.sobject = sobject
+        self.operation = operation
+        self.api_options = api_options
+        self.context = context
+        self.bulk = context.bulk
+        self.sf = context.sf
+        self.logger = context.logger
+        self.status = None
+
+
+class QueryStep(Step):
+    def __init__(self, sobject, api_options, context, query):
+        super().__init__(sobject, Operation.QUERY, api_options, context)
+        self.soql = query
+
+    def query(self):
+        pass
+
+    def get_results(self):
+        pass
+
+
+class BulkApiQueryStep(QueryStep, BulkJobTaskMixin):
+    def query(self):
+        self.job_id = self.bulk.create_query_job(self.sobject, contentType="CSV")
+        self.batch_id = self.bulk.query(self.job_id, self.soql)
+
+        result = self._wait_for_job(self.job_id)
+        if result == "Completed":
+            self.status = Status.SUCCESS
+        else:
+            self.status = Status.FAILURE
+
+        self.bulk.close_job(self.job_id)
+
+    def get_results(self):
+        # FIXME: For PK Chunking, need to get new batch Ids
+        # and retrieve their results. Original batch will not be processed.
+
+        result_ids = self.bulk.get_query_batch_result_ids(
+            self.batch_id, job_id=self.job_id
+        )
+        for result_id in result_ids:
+            uri = f"{self.bulk.endpoint}/job/{self.job_id}/batch/{self.batch_id}/result/{result_id}"
+
+            with download_file(uri, self.bulk) as f:
+                reader = csv.reader(f)
+                self.headers = next(reader)
+                if "Records not found for this query" in self.headers:
+                    return
+
+                yield from reader
+
+
+class DmlStep(Step):
+    def __init__(self, sobject, operation, api_options, context, fields):
+        super().__init__(sobject, operation, api_options, context)
+        self.fields = fields
+
+    def start(self):
+        pass
+
+    def load_records(self, records):
+        pass
+
+    def end(self):
+        pass
+
+    def get_results(self):
+        return []
+
+
+class BulkApiDmlStep(DmlStep, BulkJobTaskMixin):
+    def start(self):
+        self.job_id = self.bulk.create_job(
+            self.sobject,
+            self.operation.value,
+            contentType="CSV",
+            concurrency=self.api_options.get("bulk_mode", "Parallel"),
+        )
+
+    def end(self):
+        self.bulk.close_job(self.job_id)
+        result = self._wait_for_job(self.job_id)
+        if result == "Completed":
+            self.status = Status.SUCCESS
+        else:
+            self.status = Status.FAILURE
+
+    def load_records(self, records):
+        self.batch_ids = []
+
+        for count, batch_file in enumerate(self._batch(records)):
```

Comment:
It's not really a batch_file now; it's a generator of CSV lines, right?

Reply:
Correct. Renaming for clarity.